{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4465c508-be4e-4289-834e-47965bf9865d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import pysam \n",
    "\n",
    "f = '../../results/align/minimap2/ONT_sample-1.bam'\n",
    "f2 = 'tmp.bam'\n",
    "sam1 = pysam.AlignmentFile(f)\n",
    "sam2 = pysam.AlignmentFile(f2, \"wb\", template=sam1)\n",
    "\n",
    "i = 0\n",
    "for read in sam1:\n",
    "    i += 1\n",
    "    if i > 10:\n",
    "        break\n",
    "    sam2.write(read)\n",
    "#     break\n",
    "    \n",
    "    \n",
    "sam2.close()\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10e771d9-b40d-4663-9c2a-1a336bce51e6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# import re\n",
    "# c = read.cigarstring\n",
    "# c = re.sub('[0-9]+', '', c)\n",
    "# set(list(c))\n",
    "\n",
    "from hiseq.utils.utils import update_obj, log\n",
    "\n",
    "class Cigar(object):\n",
    "    \"\"\"\n",
    "    x : int, 0-8\n",
    "    \n",
    "    cigar table:\n",
    "    \n",
    "    BAM  Op  query  ref\n",
    "    0    M   yes    yes\n",
    "    1    I   yes    no\n",
    "    2    D   no     yes\n",
    "    3    N   no     yes\n",
    "    4    S   yes    no \n",
    "    5    H   no     no \n",
    "    6    P   no     no \n",
    "    7    =   yes    yes\n",
    "    8    X   yes    yes\n",
    "    \n",
    "    # convert CIGAR BAM to Op\n",
    "    \"\"\"\n",
    "    def __init__(self, x):\n",
    "        self.x = self.to_str(x)\n",
    "        self.on_ref = self.x in 'MDN=X'\n",
    "        self.on_query = self.x in 'MIS=X'\n",
    "        \n",
    "        \n",
    "    def to_str(self, x):\n",
    "        \"\"\"\n",
    "        MIDNSHP=X \n",
    "        012345678\n",
    "        \"\"\"\n",
    "        out = None\n",
    "        if isinstance(x, str):\n",
    "            if x in 'MIDNSHP=X':\n",
    "                out = x\n",
    "        elif isinstance(x, int):\n",
    "            if x in range(9):\n",
    "                out = 'MIDNSHP=X'[x]\n",
    "        return out\n",
    "\n",
    "\n",
    "class GetIns(object):\n",
    "    \"\"\"\n",
    "    Fetch insertions from alignment, pysam.AlignedSegment\n",
    "    \"\"\"\n",
    "    def __init__(self, x, **kwargs):\n",
    "        self = update_obj(self, kwargs, force=True)\n",
    "        self.x = x\n",
    "        # default arguments\n",
    "        args = {\n",
    "            'min_size': 300,\n",
    "            'max_size': 14000\n",
    "        }\n",
    "        self = update_obj(self, args, force=False)\n",
    "        \n",
    "        \n",
    "    def get_ins(self):\n",
    "        \"\"\"\n",
    "        a is CIGAR (tuple format) \n",
    "        \"\"\"\n",
    "        c_ref = self.x.reference_start\n",
    "        c_query = 0\n",
    "        out = [] # output\n",
    "        for a,b in self.x.cigar:\n",
    "            if a == 1 and b > self.min_size and b < self.max_size:\n",
    "                # insertion: on query, not on ref\n",
    "                # export TE insertion\n",
    "                ## coordinates on ref\n",
    "                strand = '-' if read.is_reverse else '+'\n",
    "                bed = [\n",
    "                    read.reference_name, \n",
    "                    c_ref, \n",
    "                    c_ref + 1, \n",
    "                    '{}:{}'.format(read.reference_name, c_ref),\n",
    "                    255, \n",
    "                    strand\n",
    "                ]\n",
    "                ## sequences on query\n",
    "                q_s = c_query\n",
    "                q_e = q_s + b\n",
    "                if strand == '+':\n",
    "                    ins = read.query_sequence[q_s:q_e]\n",
    "                else:\n",
    "                    s = read.query_sequence[::-1]\n",
    "                    ins = s[q_s:q_e]\n",
    "                    ins = s[::-1]\n",
    "                ## output\n",
    "                bed = bed + [b, ins] \n",
    "                bed = list(map(str, bed))\n",
    "                out.append('\\t'.join(bed))\n",
    "                ## update \n",
    "                c_query += b\n",
    "            else:\n",
    "                cigar = Cigar(a)\n",
    "                if cigar.on_ref:\n",
    "                    c_ref += b\n",
    "                if cigar.on_query:\n",
    "                    c_query += b\n",
    "        return out\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98c4d16f-e056-42d1-bd14-cd3aad22ac6d",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "# read.cigar\n",
    "\n",
    "# extract Insertion\n",
    "# criteria\n",
    "# length: 372 to 13424 \n",
    "# 300 to 14000\n",
    "\n",
    "\n",
    "sam = pysam.AlignmentFile('tmp.bam')\n",
    "\n",
    "with open('tmp.ins.bed', 'wt') as w:\n",
    "    for read in sam:\n",
    "        ins = GetIns(read).get_ins()\n",
    "        if len(ins) > 0:\n",
    "            w.write('\\n'.join(ins))\n",
    "    \n",
    "sam.close()\n",
    "\n",
    "\n",
    "\n",
    "# pos_ref = read.reference_start\n",
    "# pos_query = 0\n",
    "# out = []\n",
    "# for a,b in read.cigar:\n",
    "#     if a == 1 and b > 300 and b < 14000:\n",
    "#         # insertion: on query, not on ref\n",
    "#         # export TE insertion\n",
    "#         ## coordinates on ref\n",
    "#         strand = '-' if read.is_reverse else '+'\n",
    "#         bed = [\n",
    "#             read.reference_name, pos_ref, pos_ref + 1, \n",
    "#             '{}:{}'.format(read.reference_name, pos_ref), 255, strand]\n",
    "#         ## sequences on query\n",
    "#         pos_s = pos_query\n",
    "#         pos_e = pos_s + b\n",
    "#         ## for reverse-complement sequences\n",
    "#         s = read.query_sequence if strand == '+' else read.query_sequence[::-1]\n",
    "#         ins_seq = read.query_sequence[pos_s:pos_e]\n",
    "#         ## output\n",
    "#         bed = bed + [b, ins_seq] \n",
    "#         bed = list(map(str, bed))\n",
    "#         out.append('\\t'.join(bed))\n",
    "#         ## update \n",
    "#         pos_query += b\n",
    "#     else:\n",
    "#         cigar = Cigar(a)\n",
    "#         if cigar.on_ref:\n",
    "#             pos_ref += b\n",
    "#         if cigar.on_query:\n",
    "#             pos_query += b\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "684f86bb-8946-401b-bb34-a3b781402b35",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-01-27T00:17:55.956138Z",
     "iopub.status.busy": "2022-01-27T00:17:55.955701Z",
     "iopub.status.idle": "2022-01-27T00:17:56.088895Z",
     "shell.execute_reply": "2022-01-27T00:17:56.087792Z",
     "shell.execute_reply.started": "2022-01-27T00:17:55.956094Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "from Bio.Blast.Applications import NcbiblastnCommandline\n",
    "# from StringIO import StringIO\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import SeqIO, SearchIO\n",
    "\n",
    "# Create two sequence files\n",
    "seq1 = SeqRecord(Seq(\"FQTWEEFSRAAEKLYLADPMKVRVVLKYRHVDGNLCIKVTDDLVCLVYRTDQAQDVKKIEKF\"),\n",
    "                   id=\"seq1\")\n",
    "seq2 = SeqRecord(Seq(\"FQTWEEFSRAEKLYLADPMKVRVVLRYRHVDGNLCIKVTDDLICLVYRTDQAQDVKKIEKF\"),\n",
    "                   id=\"seq2\")\n",
    "SeqIO.write(seq1, \"seq1.fasta\", \"fasta\")\n",
    "SeqIO.write(seq2, \"seq2.fasta\", \"fasta\")\n",
    "\n",
    "blastx_cline = NcbiblastnCommandline(query=\"seq1.fasta\", subject=\"seq2.fasta\", evalue=0.001, outfmt=5, out=\"out.xml\")\n",
    "stdout, stderr = blastx_cline()\n",
    "result_handle = open(\"out.xml\")\n",
    "# blast_record = NCBIXML.read(result_handle)\n",
    "blast_records = NCBIXML.parse(result_handle)\n",
    "\n",
    "# for blast_record in blast_records:\n",
    "#     print(blast_record)\n",
    "blast_record = next(blast_records)\n",
    "    \n",
    "E_VALUE_THRESH = 0.04\n",
    "\n",
    "# for alignment in blast_record.alignments:\n",
    "#     for hsp in alignment.hsps:\n",
    "#         if hsp.expect < E_VALUE_THRESH:\n",
    "#             print(\"****Alignment****\")\n",
    "#             print(\"sequence:\", alignment.title)\n",
    "#             print(\"length:\", alignment.length)\n",
    "#             print(\"e value:\", hsp.expect)\n",
    "#             print(hsp.query[0:75] + \"...\")\n",
    "#             print(hsp.match[0:75] + \"...\")\n",
    "#             print(hsp.sbjct[0:75] + \"...\")\n",
    "\n",
    "blast_record.alignments"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
